File: tasks/20_Valid_Parentheses.go
package main

// https://leetcode.com/problems/valid-parentheses/
func isValid(s string) bool {
	stack := make([]rune, 0, len(s))
	brackets := map[rune]rune{
		'{': '}',
		'[': ']',
		'(': ')',
	}

	for _, bracket := range s {
		switch closing, ok := brackets[bracket]; {
		case ok:
			stack = append(stack, closing)
		case len(stack) == 0 || stack[len(stack)-1] != bracket:
			return false
		default:
			stack = stack[:len(stack)-1]
		}
	}

	return len(stack) == 0
}

File: tasks/20_Valid_Parentheses.py
# https://leetcode.com/problems/valid-parentheses/

class Solution:
    def isValid(self, s: str) -> bool:
        mapper = {
            "}": "{",
            "]": "[",
            ")": "(",
        }

        stack = []

        for br in s:
            if (br_open := mapper.get(br)) is None:
                stack.append(br)
            elif not stack or stack.pop() != br_open:
                return False

        return not stack
